### PR TITLE
JSON-API updates for 'data' and 'type' attributes. JSON-API Reader.

### PR DESCRIPTION
**Pull request by danelowe**
Two of the most significant changes recently to the JSON-API spec are:

* There must now be a top-level key named 'data' containing the primary resource(s). This is instead of the key being names according to the resource type.
* Each resource must have an attribute named 'type' in the output.

Also created a new Reader for JSON-API to help deserialize JSON-API formatted data into Yaks resource objects.

### CODE REVIEW

**Comment by plexus** [comment]
You can just push to the same branch again if you want to add some changes to a pull request. no need to close it and open a new one :)

File: yaks/spec/unit/yaks/format/json_api_spec.rb [hunk 1]
```diff
@@ -38,4 +39,92 @@
...
+    }
+    it 'should not include links' do
+      expect(format.call(resource)).to eql(
+         {data: [
+             {
+                 type: :wizards,
+             }
+         ]
+         }
+      )
```
Comment (plexus) [inline yaks/spec/unit/yaks/format/json_api_spec.rb]: minor remark. I probably would collapse this all to one line

``` ruby
expect(format.call(resource)).to eql( {data: [{type: :wizards}]} )
```

or even

``` ruby
expect(format.call(resource)).to eql data: [{type: :wizards}]
```

If the result still comes in in under 80 chars per line that's more than good enough.

File: yaks/lib/yaks/reader/json_api.rb [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+            links: []
+        )
+      end
+
+      def convert_embedded(links, linked)
+        links.flat_map do |rel, link_data|
+          # If this is a compound link, the link_data will contain either
+          # * 'type' and 'id' for a one to one
+          # * 'type' and 'ids' for a homogeneous to-many relationship
+          # * 'data' being an array where each member has 'type' and 'id' for heterogeneous
```
Comment (plexus) [inline yaks/lib/yaks/reader/json_api.rb]: Thank you for this! This complexity at the reading side is one of the things that IMO make JSON-API much too complex. Cool that you've actually gone to the trouble of catering for all valid cases. :)

**Comment by plexus** [comment]
Looks good to me! If you're done with it and once the build finishes I can hit the merge button.

**Comment by danelowe** [comment]
Yeah, all done for these changes. I'm sure there will be more as I come across other parts of the JSON-API specification. Thanks for the clarification on PRs!

**Comment by plexus** [comment]
thankyou

**Pull request merged**